### ISSUE TITLE
System hang when trig Memtest (Bugzilla Bug 2954)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#9979]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 2954
Date: 2020-09-03T10:27:44+00:00
From: @ydong10
To:   @ydong10
CC:   @jyao1

Last updated: 2021-07-27T05:05:38+00:00

**Comment by tianocore-issues** [comment]
## Comment 13612
Date: 2020-09-03 10:27:44 +0000
From: @ydong10

- Industry Specification: ---
- Target OS: ---
- Bugzilla Assignee(s): @ydong10

If change CR3/GDT location to above 4G range and trig StartUpThisAp to let AP do some procedure, system will hang.

Related mail discussion in edk2-devel with title:
[edk2-devel] [PATCH] UefiCpuPkg/MpInitLib: Add check for CR3/GDT/IDT.

**Comment by tianocore-issues** [comment]
## Comment 13615
Date: 2020-09-03 22:06:37 +0000
From: @ydong10

Confirmed this is an issue and provide patch for it now.

**Comment by tianocore-issues** [comment]
## Comment 13617
Date: 2020-09-03 23:22:29 +0000
From: @jyao1

I don&apos;t treat it as a valid test case, to move page table, GDT, IDT in UEFI environment.

The CPU context should be managed by CPU driver, and only CPU driver. Other driver or application should only change the system state via UEFI/PI spec defined API. Otherwise, the system may run into an unknown state.

IMHO, I dont think it is a valid case.

I suggest to make things simple to just say: dont change.

If we really treat it as valid, I suggest we define clearlly.
What we allow an driver/application to change freely?
What we do not allow a driver/application to change freely?
And what is the business justification for the decision.

**Comment by tianocore-issues** [comment]
## Comment 13618
Date: 2020-09-03 23:31:22 +0000
From: @jyao1

We also need the requirement on the expected behavior, if the expected change happens.

I doubt the value to return error during AP wake up.



Usually, AP wake up mean there is some important think we need AP to do, or we need collect information from AP.

If AP cannot be waken up, that means system might in a critical state and it may need reset.

Like in windows OS, if BSP detects AP lost, then BSOD.

**Comment by tianocore-issues** [comment]
## Comment 13619
Date: 2020-09-04 03:14:34 +0000
From: @lersek

(In reply to jiewen.yao from comment #2)
&gt; I don&apos;t treat it as a valid test case, to move page table, GDT, IDT in UEFI
&gt; environment.
&gt; 
&gt; The CPU context should be managed by CPU driver, and only CPU driver. Other
&gt; driver or application should only change the system state via UEFI/PI spec
&gt; defined API. Otherwise, the system may run into an unknown state.
&gt; 
&gt; IMHO, I dont think it is a valid case.

Agreed.

&gt; I suggest to make things simple to just say: dont change.
&gt; 
&gt; If we really treat it as valid, I suggest we define clearlly.
&gt; What we allow an driver/application to change freely?
&gt; What we do not allow a driver/application to change freely?

I don&apos;t think there&apos;s a well laid-out *technical* justification behind this use case. So I wouldn&apos;t recommend making up new rules to cover this. I&apos;m fine handling this as a one-off exception, as long as the impact on code complexity, and wasted CPU cycles, are minimal / nonexistent.

&gt; And what is the business justification for the decision.

On the other hand: yes, learning more about the business justification would be nice.

To summarize: I don&apos;t recommend that we say &quot;this application does this and that, so now we make new rules, and from this day on, *every* application can expect doing this and that&quot;. That would be really bad, IMO. Instead, I propose &quot;let&apos;s accommodate this one exceptional case, because it&apos;s really important, due to &lt;insert business justification here&gt;&quot;.

Adding new rules would complicate matters for everybody. Adding an *isolated* hack is less intrusive, IMO.
